docs/design/sum_types.md

```diff
@@ -0,0 +1,197 @@
+# Sum types
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Overview](#overview)
+-   [`choice` declarations](#choice-declarations)
+-   [User-defined sum types](#user-defined-sum-types)
+-   [Alternatives considered](#alternatives-considered)
+-   [References](#references)
+
+<!-- tocstop -->
+
+## Overview
+
+In Carbon, a _sum type_ is a type whose values are grouped into several distinct
+named cases, called _alternatives_. A value of a sum type logically consists of
+a value of one of the alternatives, together with a _discriminator_ tag, which
+identifies which alternative is present. Sum types are typically handled with
+pattern matching.
+
+## `choice` declarations
+
+The `choice` keyword is used to declare a sum type by specifying its interface,
+leaving the implementation to the compiler. For example:
+
+```carbon
+choice OptionalI32 {
+  Some(value: i32),
+  None
+}
+```
+
+This declares a sum type named `OptionalI32` with two alternatives: `Some`,
+which holds a single `i32` value, and `None`, which is empty. Choice types can
+also be parameterized, like class types:
+
+```carbon
+choice Optional(T:! Type) {
+  Some(value: T),
+  None
+}
+```
+
+An alternative declaration consists of the alternative name, followed by an
```

Comment:
Consider moving this explanation up. In other words, rather than giving examples first, explain the syntax first.